infrastructure/index.ts

```diff
@@ -267,20 +261,65 @@ crawlDirectory(
     webContentsRootPath,
     (filePath: string) => {
         const relativeFilePath = filePath.replace(webContentsRootPath + "/", "");
+
+        // If the file is a .html file that contains a meta refresh redirect, configure the bucket object to be a
+        // webpage redirect to make S3 return a 301 response when the object is requested. This is done to allow
+        // #anchors in URLs to flow across to the destination page, which browsers will do with 301 redirects, but won't
+        // do with meta refresh redirects.
+        // https://docs.aws.amazon.com/AmazonS3/latest/dev/how-to-page-redirect.html
+        const redirect = getMetaRefreshRedirect(filePath);
+
         const contentFile = new aws.s3.BucketObject(
             relativeFilePath,
             {
                 acl: "public-read",
                 key: relativeFilePath,
                 bucket: contentBucket,
-                source: new pulumi.asset.FileAsset(filePath),
-                contentType: getMimeType(filePath) || undefined,
+
+                // Use "/dev/null" for an empty object if this is a redirect, otherwise the actual file path.
+                source: new pulumi.asset.FileAsset(redirect ? "/dev/null" : filePath),
+
+                // Add `contentType` if this isn't a redirect.
+                ...!redirect && { contentType: getMimeType(filePath) },
+
+                // Add `websiteRedirect` if this is a redirect.
+                ...redirect && { websiteRedirect: redirect },
             },
             {
                 parent: contentBucket,
             });
     });
 
+// Returns the redirect URL if filePath is a .html file that contains a meta refresh redirect, otherwise undefined.
+function getMetaRefreshRedirect(filePath: string): string | undefined {
+    // Only .html files contain meta refresh redirects.
+    if (path.extname(filePath) !== ".html") {
+        return undefined;
+    }
+
+    // Extract the redirect from the content of the file.
+    const text = fs.readFileSync(filePath, "utf8");
+    const regex = /<meta\s+?http-equiv="refresh"\s+?content="0;\s+?url=(.*?)"/gm;
+    const match = regex.exec(text);
+    if (!match || match.length !== 2 || !match[1]) {
+        return undefined;
+    }
+
+    const redirect = match[1];
+
+    // If the redirect already has the https or http protocol specified, return it.
+    if (redirect.startsWith("https://") || redirect.startsWith("http://")) {
+        return redirect;
+    }
+
+    // If the redirect starts with "/", prefix with the redirect domain and return it.
+    if (redirect.startsWith("/")) {
+        return `https://${redirectDomain}${redirect}`;
+    }
+
+    return undefined;
```

Comment:
This doesn't seem right... the HTML file's redirect could exist, but just be malformed? Like "redirect='404.html'"?

Rather than returning `undefined` in this case, why don't we just return a runtime error? (Refusing to perform the Pulumi update.) That would force us to find/fix the issue, rather than not creating the redirect.

What do you think?

Reply:
Yeah, I was kind of putting my head in the sand because if there is a problem with the redirect URL, even without this change, we'd already have the problem and wouldn't know about it. But as long as we're now going through the effort to extract and use it in a different way, you're right, we should fail if it is malformed so we can fix the problem. Fixed.